util/logutil/inconsist/reporter.go

```diff
@@ -0,0 +1,264 @@
+	if !ok {
+		return ""
+	}
+	h := helper.NewHelper(tikvStore)
+	data, err := h.GetMvccByEncodedKey(key)
+	if err != nil {
+		return ""
+	}
+	regionID := getRegionIDByKey(tikvStore, key)
+
+	decodeKey := strings.ToUpper(hex.EncodeToString(key))
+
+	resp := map[string]interface{}{
+		"key":      decodeKey,
+		"regionID": regionID,
+		"mvcc":     data,
+	}
+
+	if decodeMvccFn != nil {
+		decodeMvccFn(key, data, resp)
+	}
+
+	rj, err := json.Marshal(resp)
+	if err != nil {
+		return ""
+	}
+	return string(rj)
+}
+
+func getRegionIDByKey(tikvStore helper.Storage, encodedKey []byte) uint64 {
+	keyLocation, err := tikvStore.GetRegionCache().LocateKey(tikv.NewBackofferWithVars(context.Background(), 500, nil), encodedKey)
+	if err != nil {
+		return 0
+	}
+	return keyLocation.Region.GetID()
+}
+
+// Reporter is a helper to generate report.
+type Reporter struct {
+	HandleEncode func(handle kv.Handle) kv.Key
+	IndexEncode  func(idxRow *RecordData) kv.Key
+	Tbl          *model.TableInfo
+	Idx          *model.IndexInfo
+	Sctx         sessionctx.Context
+}
+
+func (r *Reporter) decodeRowMvccData(_ kv.Key, respValue *kvrpcpb.MvccGetByKeyResponse, outMap map[string]interface{}) {
+	colMap := make(map[int64]*types.FieldType, 3)
+	for _, col := range r.Tbl.Columns {
+		colMap[col.ID] = &col.FieldType
+	}
+
+	if respValue.Info != nil {
+		var err error
+		datas := make(map[string][]map[string]string)
+		for _, w := range respValue.Info.Writes {
+			if len(w.ShortValue) > 0 {
+				datas[strconv.FormatUint(w.StartTs, 10)], err = decodeMvccData(w.ShortValue, colMap, r.Tbl)
+			}
+		}
+
+		for _, v := range respValue.Info.Values {
+			if len(v.Value) > 0 {
+				datas[strconv.FormatUint(v.StartTs, 10)], err = decodeMvccData(v.Value, colMap, r.Tbl)
+			}
+		}
+		if len(datas) > 0 {
+			outMap["decoded"] = datas
+			if err != nil {
+				outMap["decode_error"] = err.Error()
+			}
+		}
+	}
+}
+
+func (r *Reporter) decodeIndexMvccData(key kv.Key, respValue *kvrpcpb.MvccGetByKeyResponse, outMap map[string]interface{}) {
+	if respValue.Info != nil {
+		var (
+			hd    kv.Handle
+			err   error
+			datas = make(map[string]map[string]string)
+		)
+		for _, w := range respValue.Info.Writes {
+			if len(w.ShortValue) > 0 {
+				hd, err = tablecodec.DecodeIndexHandle(key, w.ShortValue, len(r.Idx.Columns))
+				if err == nil {
+					datas[strconv.FormatUint(w.StartTs, 10)] = map[string]string{"handle": hd.String()}
+				}
+			}
+		}
+		for _, v := range respValue.Info.Values {
+			if len(v.Value) > 0 {
+				hd, err = tablecodec.DecodeIndexHandle(key, v.Value, len(r.Idx.Columns))
+				if err == nil {
+					datas[strconv.FormatUint(v.StartTs, 10)] = map[string]string{"handle": hd.String()}
+				}
+			}
+		}
+		if len(datas) > 0 {
+			outMap["decoded"] = datas
+			if err != nil {
+				outMap["decode_error"] = err.Error()
+			}
+		}
+	}
+}
+
+func decodeMvccData(bs []byte, colMap map[int64]*types.FieldType, tb *model.TableInfo) ([]map[string]string, error) {
```

Comment:
Why return a `[]map[string]string` but not a `map[string]string`?

Reply:
good catch, it should be `map[string]string`